Add EDOT CF for Azure docs (#459)
elastic/opentelemetry: docs/reference/edot-cloud-forwarder/azure.md

---
navigation_title: Azure
description: Set up the EDOT Cloud Forwarder for Azure to bring your Azure logs to Elastic Observability.
applies_to:
  serverless:
    observability: preview
  product:
    edot_cf_azure: preview
products:
  - id: cloud-serverless
  - id: observability
  - id: edot-cf
---

# EDOT Cloud Forwarder for Azure

{{edot-cf}} for Azure is an agentless solution that allows you to stream logs and metrics from Azure services to an {{es}} cluster. {{edot-cf}} for Azure supports the following log and metricsources:

| Log source | Description |
| --- | --- |
| Activity log | Logs generated by Azure Monitor |
| Storage account metrics | Metrics generated by an Azure Storage |

Read on to learn how to set up {{edot-cf}} for Azure.

::::{note}
We are working to support other sources. Get in touch to let us know of any specific requirements that could influence our plans.
::::

## Architecture

> % Add notes on architecture

Reviewer: Cool. 

Do you suggest using text only, or also adding a diagram? What level of detail should I aim for?

I have several diagrams in Excalidraw like this one (PNG version):

<img width="5766" height="2768" alt="image" src="https://github.com/user-attachments/assets/912fcf1a-94d2-4c29-af82-2c73656bb30c" />

## Prerequisites

::::{important}
{{edot-cf}} for Azure is not currently supported on {{ech}} or self-managed {{stack}}.
::::

To collect logs and metrics using {{edot-cf}} for Azure, you need:

- An Azure subscription
- Azure CLI
- Bicep extension for Azure CLI
- Managed OTLP endpoint on Elastic Cloud

### Azure subscription

To use {{edot-cf}} for Azure, you need an Azure subscription with permissions to create resources.

{{edot-cf}} for Azure creates the following resources:

- Function App
- App Service plan
- Event Hubs Namespace with two Event Hubs (logs and metrics)
- Storage account
- Application Insights workspace

### Azure CLI

{{edot-cf}} for Azure uses the Azure CLI to deploy and manage resources.

To install Azure CLI, refer to the [official documentation](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli).

> ### Bicep extension

Reviewer: Um, I'm still new to Bicep, but it feels like it's an ecosystem that encourages you to use the latest version as much as possible.

Should we keep this up to date with the version we used for the last version of ECF? I'll also double-check what Bicep docs recommends.

:::{include} ../_snippets/serverless-endpoint-api.md
:::

In the Bicep templates, the OTLP endpoint is set as `elasticsearchOtlpEndpoint`, and the API key is set as `elasticsearchApiKey`. 

:::{important}
Trim the API key from `Authorization=ApiKey MYKEYVALUE...` to just `MYKEYVALUE...` before using it as the argument to the `elasticsearchApiKey` parameter.
:::
::::

## Deployment considerations

Before deploying {{edot-cf}} for Azure, keep these points in mind:

% Complete this section

## Deploy EDOT CF for Azure

Follow these steps to deploy {{edot-cf}} for Azure in a resource group to send logs and metrics to the {{motlp}} endpoint.

::::::{stepper}

:::::{step} Set the environment variables
Define the following environment variables:

```sh
export AZURE_LOCATION="<your_region>"
export RESOURCE_GROUP="<group_name>"
export ELASTICSEARCH_OTLP_ENDPOINT="<your_otlp_endpoint>"
export ELASTICSEARCH_API_KEY="<your_api_key>"
```
:::::

:::::{step} Create the resource group
Create the resource group that hosts all the resources for {{edot-cf}} for Azure.

```sh
$ az group create --name $RESOURCE_GROUP --location $AZURE_LOCATION
{

>   "id": "/subscriptions/12cabcb4-86e8-404f-a3d2-1dc9982f45ca/resourceGroups/<group_name>",

Author: Thanks!

  "location": "<your_region>",
  "managedBy": null,
  "name": "<group_name>",
  "properties": {
    "provisioningState": "Succeeded"
  },
  "tags": null,
  "type": "Microsoft.Resources/resourceGroups"
}
```
:::::

:::::{step} Download the template
Download the Bicep template to deploy {{edot-cf}} for Azure at the following URL: 

> * https://edotcfazure5gdoxpg7d2rim.blob.core.windows.net/releases/v0.6.0/ecf.bicep

Reviewer: Yeah, the current URL is pretty ugly.

The release process is not finalized yet, we have a GH issue for it, so we have room to build a better solution.

Ideally, user should use the bicep file that comes with the specific ECF version. For example, if you want to deploy ECF 0.6.0, you should use the https://edotcfazure5gdoxpg7d2rim.blob.core.windows.net/releases/v0.6.0/ecf.bicep file.

How about this: since we can omit `version=0.6.0` parameter in the `az` command, users could get the latest bicep using something like https://ela.st/edot-cf-azure-template, WDYT?

```sh
$ wget https://ela.st/edot-cf-azure-template
--2025-09-24 17:22:06--  https://ela.st/edot-cf-azure-template
Risoluzione di ela.st (ela.st)... 34.69.101.247
Connessione a ela.st (ela.st)|34.69.101.247|:443... connesso.
Richiesta HTTP inviata, in attesa di risposta... 302 Found
Posizione: https://edotcfazure5gdoxpg7d2rim.blob.core.windows.net/releases/v0.6.0/ecf.bicep [segue]
--2025-09-24 17:22:06--  https://edotcfazure5gdoxpg7d2rim.blob.core.windows.net/releases/v0.6.0/ecf.bicep
Risoluzione di edotcfazure5gdoxpg7d2rim.blob.core.windows.net (edotcfazure5gdoxpg7d2rim.blob.core.windows.net)... 20.209.179.2
Connessione a edotcfazure5gdoxpg7d2rim.blob.core.windows.net (edotcfazure5gdoxpg7d2rim.blob.core.windows.net)|20.209.179.2|:443... connesso.
Richiesta HTTP inviata, in attesa di risposta... 200 OK
Lunghezza: 7382 (7,2K) [application/octet-stream]
Salvataggio in: «edot-cf-azure-template»

edot-cf-azure-template                       100%[============================================================================================>]   7,21K  --.-KB/s    in 0s

2025-09-24 17:22:07 (1,72 GB/s) - «edot-cf-azure-template» salvato [7382/7382]
```

If you think this can fly, we need to remove `version=0.6.0` from the `az` command so it will pick the latest version from the template as we release new ECF versions.

> * https://edotcfazure5gdoxpg7d2rim.blob.core.windows.net/releases/v0.6.0/ecf.bicep

Author: Changed to https://ela.st/edot-cf-azure-template

  - In **Diagnostic setting name**, enter a name for the diagnostic setting.
  - Select all categories in **Logs > Administrative**.
  - In **Destination details**, select **Stream to an event hub**.
  - In **Event hub namespace**, select the namespace created by the EDOT Cloud Forwarder (ECF) for Azure.
  - In **Event hub name (optional)**, select the `logs` event hub created by the EDOT Cloud Forwarder (ECF) for Azure.
4. Select **Save** to create the diagnostic setting.
5. After a few minutes, the diagnostic setting starts streaming the Activity logs to the `logs` event hub.
6. Go to **Discover** in your {{ecloud}} deployment or Serverless project, and select the `logs-*` data view.
7. Filter the docs by `data_stream.dataset`, to browse the Activity logs streamed to {{es}}.

The `data_stream.dataset` value depends on your configured OTLP endpoint.

| OTLP endpoint      | Type | data_stream.dataset |
| ------------------ | ---- | ------------------- |
| Managed OTLP input | logs | `generic.otel`      |
:::

:::{tab-item} Storage Account metrics
Create a Data Collection rule to collect metrics from Storage Account metrics.

> 1. In the Azure portal, open **Data collection rules** from the search bar.

Reviewer: I'm not sure yet if setting up DCR should be part of ECF. DCR has many options (resource IDs, metric names, timegrain, etc) an maybe we should keep this separated from the ECF core.

I was considering starting with a independent bicep template to set up DCR for users interested in collecting metrics. With the option to merge it in the main bicep template if it makes sense.

WDYT @theletterf and @mlunadia?

> 1. In the Azure portal, open **Data collection rules** from the search bar.

Author: A subtemplate? That'd help abstract away these details. +1

4. In **Resources**:
  - Select **Add resources**, then select the Storage Account from which you want to collect metrics.
5. In **Collect and deliver**:
  - Select **Add new dataflow**.
  - From **Data source type**, select **Platform metrics**.
  - From **Resource types**, select the types of resources from which you want to collect metrics.
  - Select **Destination**.
  - From **Destination type**, select **Event hub**, and then select the `metrics` Event Hub created by the {{edot-cf}} for Azure.
  - Go to **Review and create** and select **Create**.
6. After some time, the Data Collection Rule start streaming the Storage Account metrics to the `metrics` event hub.
7. Go to **Discover** in your {{ecloud}} deployment or Serverless project, and select the `metrics-*` data view.
8. Filter the docs by `data_stream.dataset`, to browse the Storage Account metrics streamed to {{es}}.

The `data_stream.dataset` value depends on your configured OTLP endpoint.

| OTLP endpoint      | Type    | data_stream.dataset |
| ------------------ | ------- | ------------------- |

> | Managed OTLP input | metrics | `generic.otel`      |

Reviewer: I guess we can remove the tables now. 

I used the table when there we two possible OTLP endpoints: mOTLP and APM server. However, we're only targeting mOTLP now.

:::{note}
The Data Collection rule can take up to 30 minutes, or more, before starting to stream metrics to the Event Hub.
:::
:::
::::
:::::
::::::

## Configure the template

{{edot-cf}} for Azure uses a Bicep template to deploy the EDOT Collector.

> The following parameters are available in the Bicep template:

Reviewer: Um, let me check v0.6.0, we may have more options now. 

There will be more options when we'll merge the bicep template with support for the Flex Consumption hosting plan.

| `logsDecoder`                    | `ds`    | Decoder to decode the logs.                                          | `ds` (diagnostic settings)                                             |
| `metricsDecoder`                 | `ds`    | Decoder to decode the metrics.                                       | `ds` (diagnostic settings) and `dcr` (Data Collection Rules)           |
| `eventHubPartitionCount`         | `4`     | Number of partition in the event hubs.                               |                                                                        |
| `eventHubMessageRetentionInDays` | `1`     | Number of days for the event hub data retention.                     |                                                                        |
| `version`                        | `0.6.0` | EDOT Cloud Forwarder (ECF) for Azure Tech Preview version to deploy. | The `infra/ecf.bicep` contains the latest version available.           |

## Resource group resources

The Bicep template creates a number of resources to process logs from a specific log source.

> % Complete this section

Reviewer: Thanks for the placeholder, I'll add the resource list here.

> % Complete this section

Reviewer: Here's the list:

```suggestion
This is a list of resources created by the Bicep template.

| Resource name         | Type                                      | Description                                                                                              |
| --------------------- | ----------------------------------------- | -------------------------------------------------------------------------------------------------------- |
| `eventHubNamespace`   | `Microsoft.EventHub/namespaces`           | Event Hubs namespace to host the Event Hubs to stream logs and metrics.                                  |
| `logsEventHub`        | `Microsoft.EventHub/namespaces/eventhubs` | Event Hub for the logs.                                                                                  |
| `metricsEventHub`     | `Microsoft.EventHub/namespaces/eventhubs` | Event Hub for the metrics.                                                                               |
| `storageAccount`      | `Microsoft.Storage/storageAccounts`       | Storage Account container to store the Event Hub trigger checkpoint information and the failed messages. |
| `hostingPlan`         | `Microsoft.Web/serverfarms`               | Hosting plan that runs the function app.                                                                 |
| `functionApp`         | `Microsoft.Web/sites`                     | The function app that runs the OTel Collector.                                                           |
| `applicationInsights` | `Microsoft.Insights/components`           | Application Insights instance to monitor the application.                                                |

The  `eventHubNamespace` hosts the two event hubs responsible for receiving logs and metrics from Azure services. The `storageAccount` plays a supporting role: it stores the checkpoint information for the Event Hub triggers, and it also hosts the failed logs and metrics in the `logs-error-container` and `metrics-error-container` containers. The `functionApp` is the core component responsible for running the OTel collector inside the Azure Function. 

```

Note: strictly speaking, the `logsEventHub` and `metricsEventHub` are not subscription-level resources:

```sh
$ az resource list --resource-group mbranca-ecf-flex-2025-09-24 -o table
Name                ResourceGroup                Location    Type                               Status
------------------  ---------------------------  ----------  ---------------------------------  --------
ecfnhh77nexi2w6g    mbranca-ecf-flex-2025-09-24  eastus2     Microsoft.EventHub/namespaces
nhh77nexi2w6gecf    mbranca-ecf-flex-2025-09-24  eastus2     Microsoft.Storage/storageAccounts
nhh77nexi2w6g-plan  mbranca-ecf-flex-2025-09-24  eastus2     Microsoft.Web/serverFarms
ecfnhh77nexi2w6g    mbranca-ecf-flex-2025-09-24  eastus2     Microsoft.Insights/components
nhh77nexi2w6g-app   mbranca-ecf-flex-2025-09-24  eastus2     Microsoft.Web/sites
```

However, I think including them helps users better identify the most important components.

---
navigation_title: Azure
description: Set up the EDOT Cloud Forwarder for Azure to bring your Azure logs to Elastic Observability.
applies_to:
  serverless:
    observability: preview
  product:
    edot_cf_azure: preview
products:
  - id: cloud-serverless
  - id: observability
  - id: edot-cf
---

# EDOT Cloud Forwarder for Azure

> {{edot-cf}} for Azure is an agentless solution that allows you to stream logs and metrics from Azure services to an {{es}} cluster. {{edot-cf}} for Azure supports the following log and metricsources:

Reviewer 2: ```suggestion
{{edot-cf}} for Azure is an serverless function based solution that allows you to stream logs and metrics from Azure services to an {{es}} cluster. {{edot-cf}} for Azure supports the following log types:
```

| Source | Description |
| --- | --- |
| Activity log | Logs generated by Azure Monitor |

> | Storage account metrics | Metrics generated by an Azure Storage |

Reviewer 2: @zmoog I don't think we have discussed metrics support.

Reviewer: The original ECF PoC included metrics support.

ECF has basic support for Diagnostic Settings (DS) and Data Collection Rules (DCR) metrics.

By "basic support," I mean that metrics in the DS and DCR formats land in the `metrics-generic.otel-default` data stream.

For the DS metrics format, we're using the upstream translator. For the DCR metrics format, we're using a new translator.

We can omit metrics and focus on logs only if you'd like to review the current metrics support first.

Read on to learn how to set up {{edot-cf}} for Azure.

::::{note}
We are working to support other sources. Get in touch to let us know of any specific requirements that could influence our plans.
::::

## Architecture

% Add notes on architecture

## Prerequisites

::::{important}

> {{edot-cf}} for Azure is not currently supported on {{ech}} or self-managed {{stack}}.

Reviewer 2: I think we should change this message and also in AWS to something along the lines fo EDOT CF requires an OTLP endpoint as destination, in serverless and soon in ECH you can use mOTLP (link) and in self-managed deploy EDOT Collector in gateway mode (link)

::::

> To collect logs and metrics using {{edot-cf}} for Azure, you need:

Reviewer 2: let's remove metrics for now, until we are certain we can support them.

- An Azure subscription
- Azure CLI
- Bicep extension for Azure CLI

> - Managed OTLP endpoint on Elastic Cloud

Reviewer 2: This reads like an odd requirement, perhaps phrase it as 'getting the endpoint' and clarify that mOTLP is available in Serverless and soon on ECH, add link to instructions on how to get it.

### Azure subscription

> To use {{edot-cf}} for Azure, you need an Azure subscription with permissions to create resources.
> 
> {{edot-cf}} for Azure creates the following resources:

Reviewer 2: ```suggestion
To use {{edot-cf}} for Azure, you need an Azure subscription with permissions to create the following resources:
```

- Function App
- App Service plan
- Event Hubs Namespace with two Event Hubs (logs and metrics)
- Storage account
- Application Insights workspace

### Azure CLI

> {{edot-cf}} for Azure uses the Azure CLI to deploy and manage resources.

Reviewer 2: ```suggestion
You can use the Azure CLI to deploy and manage resources required for {{edot-cf}}.
```

To install Azure CLI, refer to the [official documentation](https://learn.microsoft.com/en-us/cli/azure/install-azure-cli).

### Bicep extension

{{edot-cf}} for Azure uses Bicep to define all the necessary resources and their dependencies.

After you've installed Azure CLI, install Bicep by running the following command:

```bash
az bicep install
```

### Managed OTLP endpoint

> {{edot-cf}} for Azure requires a {{motlp}} endpoint on {{ecloud}} and an API key.

Reviewer 2: good to clarify availability and how to procure it.

To find out the URL of the managed OTLP endpoint and the API key for authentication, follow these steps:

::::{dropdown} Steps to retrieve the OTLP endpoint and API key

> :::{include} ../_snippets/serverless-endpoint-api.md

Reviewer 2: ok spotted this after all my comments. We still pending on mentioning this is currently only for serverless and soon on ECH.

:::

In the Bicep templates, the OTLP endpoint is set as `elasticsearchOtlpEndpoint`, and the API key is set as `elasticsearchApiKey`. 

:::{important}
Trim the API key from `Authorization=ApiKey MYKEYVALUE...` to just `MYKEYVALUE...` before using it as the argument to the `elasticsearchApiKey` parameter.
:::
::::

## Deployment considerations

Before deploying {{edot-cf}} for Azure, keep these points in mind:

> Before deploying {{edot-cf}} for Azure, take the following into consideration:

Reviewer 3: I think you left 2 versions by mistake. I would go with the second one.

> - The logs event hub supports Azure resource logs, and the metrics event hub supports DCR metrics.

Reviewer: We can hold off on metrics until we have a clear e2e use case. For Storage Account metrics, ECF probably has an edge over other collection methods, so I would suggest starting from there.

- Support for additional encoding extensions for logs and metrics will be added in future releases.

## Deploy EDOT CF for Azure

Follow these steps to deploy {{edot-cf}} for Azure in a resource group to send logs and metrics to the {{motlp}} endpoint.

::::::{stepper}

:::::{step} Set the environment variables
Define the following environment variables:

```sh
export AZURE_LOCATION="<your_region>"
export RESOURCE_GROUP="<group_name>"
export ELASTICSEARCH_OTLP_ENDPOINT="<your_otlp_endpoint>"
export ELASTICSEARCH_API_KEY="<your_api_key>"
```
:::::

:::::{step} Create the resource group
Create the resource group that hosts all the resources for {{edot-cf}} for Azure.

```sh
$ az group create --name $RESOURCE_GROUP --location $AZURE_LOCATION
{
  "id": "/subscriptions/<subscription_id>/resourceGroups/<group_name>",
  "location": "<your_region>",
  "managedBy": null,
  "name": "<group_name>",
  "properties": {
    "provisioningState": "Succeeded"
  },
  "tags": null,
  "type": "Microsoft.Resources/resourceGroups"
}
```
:::::

:::::{step} Download the template

> Download the Bicep template to deploy {{edot-cf}} for Azure at the following URL: 
> 
> * https://ela.st/edot-cf-azure-template

Reviewer 3: ```suggestion
Download the Bicep template to deploy {{edot-cf}} for Azure at the following URL: https://ela.st/edot-cf-azure-template
```

Is there a point in making a list with only one bullet point?

  - Enter a **Name** for the Data Collection Rule.
  - Select or create a new **Resource group**. You can use the same resource group created by the {{edot-cf}} for Azure.
  - In **Identity**, turn on **Enable Managed Identity** and select **System assigned**. This streams the metrics to the Event Hub.
4. In **Resources**:
  - Select **Add resources**, then select the Storage Account from which you want to collect metrics.
5. In **Collect and deliver**:
  - Select **Add new dataflow**.
  - From **Data source type**, select **Platform metrics**.
  - From **Resource types**, select the types of resources from which you want to collect metrics.
  - Select **Destination**.
  - From **Destination type**, select **Event hub**, and then select the `metrics` Event Hub created by the {{edot-cf}} for Azure.
  - Go to **Review and create** and select **Create**.
6. After some time, the Data Collection Rule start streaming the Storage Account metrics to the `metrics` event hub.
7. Go to **Discover** in your {{ecloud}} deployment or Serverless project, and select the `metrics-*` data view.

> 8. Filter the docs by `data_stream.dataset`, to browse the Storage Account metrics streamed to {{es}}.

Reviewer: ECF is more scalable that Beats if users want to collect metrics with PT1M time grain.

Each Storage Account has five namespaces, each one with its own metrics, dimensions, and aggregation types. 

Scape all these data from the API every 60s is quite challenging for for Beats, even using the batch API. Beats (or the azuremonitorreceiver from OTel) can only run on a single instance.

With DCR, Azure can stream all these metrics to an event hub, and ECF can scale out to process them on multiple workers.

| --------------------- | ----------------------------------------- | -------------------------------------------------------------------------------------------------------- |
| `eventHubNamespace`   | `Microsoft.EventHub/namespaces`           | Event Hubs namespace to host the event hubs to stream logs and metrics.                                  |
| `logsEventHub`        | `Microsoft.EventHub/namespaces/eventhubs` | Event hub for the logs.                                                                                  |
| `metricsEventHub`     | `Microsoft.EventHub/namespaces/eventhubs` | Event hub for the metrics.                                                                               |
| `storageAccount`      | `Microsoft.Storage/storageAccounts`       | Storage Account container to store the Event Hubs trigger checkpoint information and the failed messages. |
| `hostingPlan`         | `Microsoft.Web/serverfarms`               | Hosting plan that runs the function app.                                                                 |

> | `functionApp`         | `Microsoft.Web/sites`                     | The function app that runs the Collector.                                                           |

Reviewer 3: ```suggestion
| `functionApp`         | `Microsoft.Web/sites`                     | The function app that runs the EDOT Collector.                                                           |
```

?

| `applicationInsights` | `Microsoft.Insights/components`           | Application Insights instance to monitor the application.                                                |

Note the following:

> - The  `eventHubNamespace` hosts the two event hubs responsible for receiving logs and metrics from Azure services. 

Reviewer 3: ```suggestion
- The `eventHubNamespace` hosts the two event hubs responsible for receiving logs and metrics from Azure services. 
```

```
:::::

:::::{step} Create the resource group
Create the resource group that hosts all the resources for {{edot-cf}} for Azure.

```sh
$ az group create --name $RESOURCE_GROUP --location $AZURE_LOCATION
{
  "id": "/subscriptions/12cabcb4-86e8-404f-a3d2-1dc9982f45ca/resourceGroups/<group_name>",
  "location": "<your_region>",
  "managedBy": null,
  "name": "<group_name>",
  "properties": {
    "provisioningState": "Succeeded"
  },
  "tags": null,
  "type": "Microsoft.Resources/resourceGroups"
}
```
:::::

:::::{step} Download the template
Download the Bicep template to deploy {{edot-cf}} for Azure at the following URL: 

* https://edotcfazure5gdoxpg7d2rim.blob.core.windows.net/releases/v0.6.0/ecf.bicep
:::::

:::::{step} Deploy the Bicep template
Deploy the Bicep template using the following command:

```sh
az deployment group create \
    --resource-group ${RESOURCE_GROUP} \
    --template-file ecf.bicep \
    --parameters \
        elasticsearchOtlpEndpoint=$ELASTICSEARCH_OTLP_ENDPOINT \
        elasticsearchApiKey=$ELASTICSEARCH_API_KEY \
        logsDecoder=ds \ <1>
        metricsDecoder=dcr \ <2>
        eventHubPartitionCount=8 \
        eventHubMessageRetentionInDays=1 \

>         version=0.6.0 <3>

Author: @zmoog So using https://ela.st/edot-cf-azure-template we can omit the version field from this command?
